### DISCUSSION TITLE
Spacing between widgets in a container

### DISCUSSION THREAD

**Discussion by aaronst** [Textualize/textual#1290] [Q&A] [answered]
I often only want `margin` _between_ widgets. Almost like `grid-gutter`, but without the need for a grid layout.

```
+-------------------+
|+-----------------+| <- no space
||     widget      ||
|+-----------------+|
|                   | <- space
|+-----------------+|
||     widget      ||
|+-----------------+| <- no space
+-------------------+
```

Typically I want `padding` to dictate the space around the content of the container, but the addition of `margin` causes that space to be larger than desired. Is there a simple way to accomplish this?

**Comment by learnbyexample** [comment]
Have you tried setting `margin-top` and `margin-bottom`? See https://textual.textualize.io/styles/margin/

![image](https://user-images.githubusercontent.com/17766317/204071570-e0b5c6c1-b431-4186-b085-5ca7f76d585b.png)

**Comment by willmcgugan** [comment, marked as answer]
No trivial way currently. You could set the margin to 0 for the first and last, but you would have to set those classes manually. We do plan on adding :first and :last pseudo selectors which would make this easier.

**Reply by aaronst** [reply]
I've been able to get away with:

```python
if self.children:
    widget.styles.margin = (1, 0, 0, 0)

self.mount(widget)
```

but `:first` and `:last` will be fantastic.

**Reply by devraj** [reply]
Wondering if this made it to the later versions of textual?

**Reply by TomJGooding** [reply]
> Wondering if this made it to the later versions of textual?

https://textual.textualize.io/guide/CSS/#pseudo-classes

**Comment by hellow554** [comment]
Sorry to resurrect this thread but I have the exact same question. I have opted for a nested horizontalgroup which seems to work?! Is there a better approach for this?

```python
    @override
    def compose(self) -> ComposeResult:
        with HorizontalGroup():
            with HorizontalGroup():
                yield Label("FooBar")
            yield Button("↻", variant="warning")
```

<img width="954" height="320" alt="grafik" src="https://github.com/user-attachments/assets/1729ba04-f960-4f5d-b0ee-8ee4eda8eb39" />

**Reply by willmcgugan** [reply]
See the `first-*` and `last-*` psuedo-selectors https://textual.textualize.io/guide/CSS/#pseudo-classes

**Reply by hellow554** [reply]
Yeah, I saw that the pseudo classes were implemented but failed to achieve something with them.
Do you have a small css snippet you could provide?
In other frameworks you have a spacer widget which consumes the left over space for you.
If you specify `expand=True` on the label, the button vanishes. The code for that only sets the width to the container width, but does not respect other widgets on that container as well...

**Reply by willmcgugan** [reply]
If you want assistance, please write a fully working example that can be run without modification. And please open a new discussion. Or post on the Discord server...
